SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -214,33 +214,59 @@ void SpatialEventTracer::StreamDeleter::operator()(Io_Stream* StreamToDestroy) c
 	Io_Stream_Destroy(StreamToDestroy);
 }
 
-void SpatialEventTracer::AddComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId, const FSpatialGDKSpanId& SpanId)
+void SpatialEventTracer::BeginOpsForFrame()
 {
-	EntityComponentSpanIds.FindOrAdd({ EntityId, ComponentId }, SpanId);
+	EntityComponentSpanIds.Empty(
+		EntityComponentSpanIds.Num()); // Reset all entries. It is assumed all entries are consumed during processing.
 }
 
-void SpatialEventTracer::RemoveComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+void SpatialEventTracer::AddEntity(const Worker_AddEntityOp& Op, const FSpatialGDKSpanId& SpanId)
 {
-	EntityComponentSpanIds.Remove({ EntityId, ComponentId });
+	TraceEvent(FSpatialTraceEventBuilder::CreateReceiveCreateEntity(Op.entity_id), SpanId.GetConstId(), 1); // No continuation
 }
 
-void SpatialEventTracer::UpdateComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId, const FSpatialGDKSpanId& SpanId)
+void SpatialEventTracer::RemoveEntity(const Worker_RemoveEntityOp& Op, const FSpatialGDKSpanId& SpanId)
 {
-	FSpatialGDKSpanId& StoredSpanId = EntityComponentSpanIds.FindChecked({ EntityId, ComponentId });
-	FSpatialGDKSpanId CauseSpanIds[2] = { SpanId, StoredSpanId };
-	StoredSpanId = TraceEvent(FSpatialTraceEventBuilder::CreateMergeComponentUpdate(EntityId, ComponentId),
-							  reinterpret_cast<uint8_t*>(&CauseSpanIds), 2);
+	TraceEvent(FSpatialTraceEventBuilder::CreateReceiveRemoveEntity(Op.entity_id), SpanId.GetConstId(), 1); // No continuation
 }
 
-FSpatialGDKSpanId SpatialEventTracer::GetSpanId(const EntityComponentId& Id) const
+void SpatialEventTracer::AuthorityChange(const Worker_ComponentSetAuthorityChangeOp& Op, const FSpatialGDKSpanId& SpanId)
 {
-	const FSpatialGDKSpanId* SpanId = EntityComponentSpanIds.Find(Id);
-	if (SpanId == nullptr)
+	TraceEvent(
+		FSpatialTraceEventBuilder::CreateAuthorityChange(Op.entity_id, Op.component_set_id, static_cast<Worker_Authority>(Op.authority)),
+		SpanId.GetConstId(), 1); // No continuation
+}
+void SpatialEventTracer::AddComponent(const Worker_AddComponentOp& Op, const FSpatialGDKSpanId& SpanId)
+{
+	if (!SpanId.IsNull())
```

Comment:
I think we should propagate the null SpanIds. I think there is a difference between the cause of an event created with a null spanId and created with no SpanId. No SpanId would mean it is a root event. A null SpanId would mean the somewhere along the causes, we decided to not sample an event.